
<!DOCTYPE html>

<html>
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <title>RIP documentation &#8212; RIP beta documentation</title>
    <link rel="stylesheet" href="_static/alabaster.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    <script id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
    <script src="_static/jquery.js"></script>
    <script src="_static/underscore.js"></script>
    <script src="_static/doctools.js"></script>
    <script src="_static/language_data.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
   
  <link rel="stylesheet" href="_static/custom.css" type="text/css" />
  
  
  <meta name="viewport" content="width=device-width, initial-scale=0.9, maximum-scale=0.9" />

  </head><body>
  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          

          <div class="body" role="main">
            
  <div class="section" id="rip-documentation">
<h1>RIP documentation<a class="headerlink" href="#rip-documentation" title="Permalink to this headline">¶</a></h1>
<div class="toctree-wrapper compound">
</div>
</div>
<div class="section" id="introduction">
<h1>Introduction<a class="headerlink" href="#introduction" title="Permalink to this headline">¶</a></h1>
<p>RIP is a comprehensive computational tool for identifying species from DNA sequence data. It can pre-process sequencing data (adapter trimming), build a database, analyse sequencing data and perform a deamination profile analysis. It can work in two different modes. One mode performs metagenomic identifications from samples containing multiple organisms, and outputs mea posterior species abundances. The second mode can perform species identifications from single organism samples and it outputs species assignment posterior probabilities.</p>
</div>
<div class="section" id="installation">
<h1>Installation<a class="headerlink" href="#installation" title="Permalink to this headline">¶</a></h1>
<p>RIP can be installed though pip, bioconda, and you can download the source from git as well.</p>
<p>Pip</p>
<blockquote>
<div><p>pip install rip</p>
</div></blockquote>
<p>Conda</p>
<blockquote>
<div><p>conda install -c bioconda rip</p>
</div></blockquote>
<dl class="simple">
<dt>Git</dt><dd><p>git clone rip</p>
</dd>
</dl>
</div>
<div class="section" id="tutorial">
<h1>Tutorial<a class="headerlink" href="#tutorial" title="Permalink to this headline">¶</a></h1>
<div class="section" id="building-the-database">
<h2>Building the database<a class="headerlink" href="#building-the-database" title="Permalink to this headline">¶</a></h2>
<p>In order to build the database we will be using the <cite>database</cite> submodule of RIP.
First we need to know what what organisms we would like to include in our database. Do we only need the complete genomes of a specific genus or do we want more genera?</p>
</div>
<div class="section" id="constructing-the-query">
<h2>Constructing the Query<a class="headerlink" href="#constructing-the-query" title="Permalink to this headline">¶</a></h2>
<p>After deciding what taxa we would like to include in our database we need to construct an NCBI query that will return all the accessions that belong to the taxa that we are interested in. One of the best ways to construct such a query is to go on the website of  NCBI’s Nucleotide database (<a class="reference external" href="https://www.ncbi.nlm.nih.gov/nucleotide/">https://www.ncbi.nlm.nih.gov/nucleotide/</a>), type in our query and get the correctly formatted query string from the “Search details” box. We can then use that string for the construction of our database.</p>
<p>For example if we would like to build a database of the complete genomes of the species of the Yersinia genus we can use the following command</p>
<blockquote>
<div><p>rip_multilevel database –query ‘(“Yersinia”[Organism] OR “Yersinia”[Organism]) AND “complete genome”[All Fields]’ –db-output ./yersinia_example</p>
</div></blockquote>
<p>For each species (or any other user defined taxonomic rank), the longest sequence per taxon will be used to populate our database.</p>
</div>
<div class="section" id="representative-refseq-species">
<h2>Representative RefSeq species<a class="headerlink" href="#representative-refseq-species" title="Permalink to this headline">¶</a></h2>
<p>When constructing a database there is always the option to include the species of the Representative RefSeq database as well. All you need to do is include the corresponding flag in your command.</p>
<blockquote>
<div><p>rip_multilevel database –query ‘(“Yersinia”[Organism] OR “Yersinia”[Organism]) AND “complete genome”[All Fields]’ –db-output ./yersinia_example –refseq-rep True</p>
</div></blockquote>
</div>
<div class="section" id="providing-custom-accessions">
<h2>Providing custom accessions<a class="headerlink" href="#providing-custom-accessions" title="Permalink to this headline">¶</a></h2>
<p>It is also possible to provide your own accessions for a selected species/taxon. For that you will need to prepare a tab delimited file with two columns. The first column is the name of the taxon, that cannot contain any special characters, other than an underscore (‘_’), and the second column is a valid NCBI accession code.</p>
<p>Here is an example of such a file:</p>
<blockquote>
<div><p>Felis_silvestris_silvestris     NC_028310.1
Felis_silvestris_lybica KP202275.1</p>
</div></blockquote>
<p>The we can simply run the following command</p>
<p>rip_multilevel database –query ‘(“Yersinia”[Organism] OR “Yersinia”[Organism]) AND “complete genome”[All Fields]’ –db-output ./yersinia_example –accessions acc_example.txt</p>
</div>
<div class="section" id="providing-custom-sequences">
<h2>Providing custom sequences<a class="headerlink" href="#providing-custom-sequences" title="Permalink to this headline">¶</a></h2>
<p>It is also possible to provide your own sequences for a taxon. To do that you will need a a tab delimited file containing the the name of the taxon with no special characters, and an underscore (‘_’) instead of spaces, a user defined accession code and the path of the fasta file. The fasta file that the path point to can be either uncompressed or compressed with gzip/bgzip.</p>
<p>Here is an example of such a file:</p>
<blockquote>
<div><p>Felis_silvestris_silvestris     user_seq_1      /path/to/fasta/file</p>
</div></blockquote>
<p>The we can simply run the following command</p>
<blockquote>
<div><p>rip_multilevel database –query ‘(“Yersinia”[Organism] OR “Yersinia”[Organism]) AND “complete genome”[All Fields]’ –db-output ./yersinia_example –sequences seq_example.txt</p>
</div></blockquote>
</div>
<div class="section" id="combinations">
<h2>Combinations<a class="headerlink" href="#combinations" title="Permalink to this headline">¶</a></h2>
<p>All of the previous options can be combined into one command. It is important to note that only one sequence per taxon is allowed in our database, and the priority goes user defined accessions or sequences, representative RefSeq and then user specified query.</p>
</div>
<div class="section" id="index-building">
<h2>Index building<a class="headerlink" href="#index-building" title="Permalink to this headline">¶</a></h2>
<p>For the first part of the analysis an index out of all the genomes that are included in our database needs to be build. This is a process that can take big amounts of memory based on the number and the complexity of the sequences that exist in our database. For that reason the user can specify the desired amount of memory resources available to rip and the program will try to build the required index. This can be specified through the <cite>–mem</cite> flag, that can be appended to the any of the commands shown above. It is important to note that the memory specified needs to be in MB.</p>
</div>
<div class="section" id="database-building-modes">
<h2>Database building modes<a class="headerlink" href="#database-building-modes" title="Permalink to this headline">¶</a></h2>
<p>For the complete construction of a database, sequences need to be downloaded and subseqeuently indexed.
The default mode of the module <cite>rip database</cite> is <cite>build</cite>, which downloads and indexes all the sequences that have been requested by the user.
If a user would like to just download the sequences and index them later it is possible to do so, by specifying the flag <cite>rip database –mode fetch</cite>, to download the sequences first and then execute <cite>rip database –mode index</cite> in order to perform the indexing.
If mode <cite>fetch</cite> is run first the mode <cite>index</cite> should be run subsequently, and not mode <cite>build</cite>, otherwise an error will be raised. If the user would like to build the database in one step then mode <cite>build</cite> is the correct choice.</p>
<p>Here is an example of building a database in two steps instead of one</p>
<blockquote>
<div><p>rip_multilevel database –mode fetch –query ‘(“Yersinia”[Organism] OR “Yersinia”[Organism]) AND “complete genome”[All Fields]’ –db-output ./yersinia_example
rip_multilevel database –mode build –db-output ./yersinia_example</p>
</div></blockquote>
</div>
<div class="section" id="preparing-a-sample-for-analysis">
<h2>Preparing a sample for analysis<a class="headerlink" href="#preparing-a-sample-for-analysis" title="Permalink to this headline">¶</a></h2>
<p>After our database is built we need to prepare our samples for analysis. In order to do that we are going to use the <cite>sample</cite> module of rip. The input files can be SE, PE or collapsed reads. If the reads are collapsed they are going to be treated as SE reads.</p>
<p>It is possible to trim sequencing adapters and collapse PE reads by specifying the relative flags, and it also possible to download samples from the SRA if an sra accession is provided.</p>
<p>If you have SE or already collapsed reads you only need to specify a file path for the <cite>–fastq</cite> flag, but if your input is PE reads then you’ll need to specify file paths for both the <cite>–fastq1</cite> and <cite>–fastq2</cite> flags. If you want to download files from the SRA all you need to do is provide an SRA accession for the <cite>–sra</cite> flag.</p>
<p>Here is an example of downloading reads from the SRA, trimming sequencing adapters and collapsing reads.</p>
<blockquote>
<div><p>rip_multilevel sample –sra SRR12157896 –collapse –sample-output-dir samples</p>
</div></blockquote>
</div>
<div class="section" id="sample-analysis">
<h2>Sample analysis<a class="headerlink" href="#sample-analysis" title="Permalink to this headline">¶</a></h2>
<p>In order to analyse any sample we will need to use the <cite>analyse</cite> module of rip.</p>
</div>
<div class="section" id="filtering-alignment">
<h2>Filtering Alignment<a class="headerlink" href="#filtering-alignment" title="Permalink to this headline">¶</a></h2>
<p>The first step for the sample analysis is to filter in all the reads that align to any of the genomes in our database. For that we will need to use the <cite>rip analyse –mode filter</cite>.</p>
<p>Here is an example command</p>
<blockquote>
<div><p>rip_multilevel analyse –mode filter –database yersinia_example/database_build_config.yaml –sample ./samples/SRR12157896_config.yaml –analysis-output-dir ./analysis_output</p>
</div></blockquote>
</div>
<div class="section" id="database-alignments">
<h2>Database Alignments<a class="headerlink" href="#database-alignments" title="Permalink to this headline">¶</a></h2>
<p>After we have filtered our libraries we can align the filtered reads against all the genomes that are included in our database. This can be done by using mode <cite>align</cite> of <cite>rip analyse</cite>.</p>
<p>For example</p>
<blockquote>
<div><p>rip_multilevel analyse –mode align –database yersinia_example/database_build_config.yaml –sample ./samples/SRR12157896_config.yaml –analysis-output-dir ./analysis_output</p>
</div></blockquote>
</div>
<div class="section" id="likelihood-calculation">
<h2>Likelihood calculation<a class="headerlink" href="#likelihood-calculation" title="Permalink to this headline">¶</a></h2>
<p>After all the metagenomic individual alignments have been competed, the number of transitions and transversions will be calculated for every read that has aligned against any of the reference genomes in our database. Then the likelihoods and posterior probabilities for each read being sampled for a given reference genome will be calculated. For this step we can use the <cite>likelihoods</cite> mode of <cite>rip analyse</cite>.</p>
<blockquote>
<div><p>rip_multilevel analyse –mode likelihoods –database yersinia_example/database_build_config.yaml –sample ./samples/SRR12157896_config.yaml –analysis-output-dir ./analysis_output</p>
</div></blockquote>
</div>
<div class="section" id="single-organism-sample-or-metagenome">
<h2>Single organism sample or metagenome ?<a class="headerlink" href="#single-organism-sample-or-metagenome" title="Permalink to this headline">¶</a></h2>
<p>Depending on whether we would like to identify the species a sample is belongs to, or perform a metagenomic analysis, we can use the <cite>probabilities</cite> or <cite>abundances</cite> mode of <cite>rip analyse</cite> respectively.</p>
</div>
<div class="section" id="assignment-probability-calculation">
<h2>Assignment Probability Calculation<a class="headerlink" href="#assignment-probability-calculation" title="Permalink to this headline">¶</a></h2>
<p>In order to calculate posterior assignment probabilities we can run the following command</p>
<blockquote>
<div><p>rip_multilevel analyse –mode probabilities –database yersinia_example/database_build_config.yaml –sample ./samples/SRR12157896_config.yaml –analysis-output-dir ./analysis_output</p>
</div></blockquote>
</div>
<div class="section" id="mean-posterior-abundances">
<h2>Mean Posterior Abundances<a class="headerlink" href="#mean-posterior-abundances" title="Permalink to this headline">¶</a></h2>
<p>In order to calculate mean posterior abundances we can run the following command</p>
<blockquote>
<div><p>rip_multilevel analyse –mode abundances –database yersinia_example/database_build_config.yaml –sample ./samples/SRR12157896_config.yaml –analysis-output-dir ./analysis_output</p>
</div></blockquote>
</div>
<div class="section" id="mapdamage-analysis">
<h2>Mapdamage analysis<a class="headerlink" href="#mapdamage-analysis" title="Permalink to this headline">¶</a></h2>
<p>If our samples are ancient we can use mapDamage to estimate the level of deamination in all the reads that have aligned to any taxon in our database. For that we can use the <cite>mapdamage</cite> module of rip.</p>
<p>Here is an example command</p>
<blockquote>
<div><p>rip_multilevel analyse –mode mapdamage –database yersinia_example/database_build_config.yaml –sample ./samples/SRR12157896_config.yaml –analysis-output-dir ./analysis_output</p>
</div></blockquote>
</div>
<div class="section" id="important-note">
<h2>Important note<a class="headerlink" href="#important-note" title="Permalink to this headline">¶</a></h2>
<p>the first 3 steps (filter, align, likelihoods) can be executed automatically when you call the probabilities or abundances mode of rip.</p>
</div>
</div>
<div class="section" id="command-line-interface">
<h1>Command Line Interface<a class="headerlink" href="#command-line-interface" title="Permalink to this headline">¶</a></h1>
<div class="section" id="rip-config">
<h2>rip config<a class="headerlink" href="#rip-config" title="Permalink to this headline">¶</a></h2>
<blockquote>
<div><dl class="option-list">
<dt><kbd><span class="option">-h</span>, <span class="option">--help</span></kbd></dt>
<dd><p>show this help message and exit</p>
</dd>
</dl>
<p>-e , –email          Email address for NCBI identification. Mandatory.
-gc , –genome-cache-folder</p>
<blockquote>
<div><p>Path where all the genomes that are downloaded and/or
used by rip are being stored. (default ~/rip_genomes/)</p>
</div></blockquote>
<dl class="simple">
<dt>-b , –batchsize      Batchsize for fetching records from NCBI &lt;int&gt;</dt><dd><p>(default: 5)</p>
</dd>
<dt>-mp , –mismatch-probability</dt><dd><p>Base mismatch probability &lt;float&gt; (default: 0.05)</p>
</dd>
<dt>-t , –bowtie2-threads</dt><dd><p>Threads for the bowtie2 alignments &lt;int&gt; (default: 1)</p>
</dd>
<dt>-s , –bowtie2-scaling</dt><dd><p>Factor to rescale/chunk the input file for the
mutlifasta index for the filtering alignment (default:
2.5)</p>
</dd>
</dl>
</div></blockquote>
</div>
<div class="section" id="rip-database">
<h2>rip database<a class="headerlink" href="#rip-database" title="Permalink to this headline">¶</a></h2>
<blockquote>
<div><dl class="option-list">
<dt><kbd><span class="option">-h</span>, <span class="option">--help</span></kbd></dt>
<dd><p>show this help message and exit</p>
</dd>
</dl>
<p>–dry-run
-m , –mode           Database creation mode for rip
-o , –db-output      /path/to/what/used/to/be/query_name
-R REFSEQ_REP, –refseq-rep REFSEQ_REP</p>
<blockquote>
<div><p>Use the prokaryotic representative species of the
RefSeq DB for the species id pipeline. only species no
strains. either or both of with_refseq_rep and
with_entrez_query should be set (default: False)</p>
</div></blockquote>
<p>-q , –query          Actual NCBI query in the NCBI query language
-Q , –query-file     Actual NCBI query in the NCBI query language
-r , –rank           Taxonomic rank to perform the identifications on</p>
<blockquote>
<div><p>(genus, species, subspecies, serotype) &lt;str&gt; (default:
species)</p>
</div></blockquote>
<dl class="simple">
<dt>-s , –sequences      TAB DELIMITED input file containing the the name of</dt><dd><p>the taxon with no special characters, and an
underscore ‘_’ instead of spaces, a user defined
accession code and the path of the fasta file. The
fasta file that the path point to can be either
uncompressed or compressed with gzip/bgzip</p>
</dd>
<dt>-a , –accessions     TAB DELIMITED input file containing the the name of</dt><dd><p>the taxon with no special characters, and an
underscore ‘_’ instead of spaces, a user defined valid
NCBI nucleotide, assembly or WGS accession code.</p>
</dd>
</dl>
<dl class="option-list">
<dt><kbd><span class="option">-g</span></kbd></dt>
<dd><p>[ …], –genera  [ …]
List containing the names of specific genera the
abundances should be calculated on, separated by a
space character &lt;genus1 genus2 genus3 …&gt;</p>
</dd>
</dl>
<p>-c , –cores          Number of cores for RIP to use
-M , –mem            Max memory resources allowed to be used ofr indexing</p>
<blockquote>
<div><p>the input for the filtering alignment (default: max
available memory 8192.0)</p>
</div></blockquote>
<dl class="option-list">
<dt><kbd><span class="option">-u</span>, <span class="option">--unlock</span></kbd></dt>
<dd><p>Unlock the working directory after smk is abruptly
killed &lt;bool&gt; (default: False)</p>
</dd>
<dt><kbd><span class="option">-d</span>, <span class="option">--debug</span></kbd></dt>
<dd><p>Debug the RIP workflow &lt;bool&gt; (default: False)</p>
</dd>
</dl>
<p>-smk , –snakemake    Snakemake flags (default: ‘’)</p>
</div></blockquote>
</div>
<div class="section" id="rip-sample">
<h2>rip sample<a class="headerlink" href="#rip-sample" title="Permalink to this headline">¶</a></h2>
<blockquote>
<div><dl class="option-list">
<dt><kbd><span class="option">-h</span>, <span class="option">--help</span></kbd></dt>
<dd><p>show this help message and exit</p>
</dd>
</dl>
<dl class="simple">
<dt>-p , –sample-prefix</dt><dd><p>Sample prefix for all the future analysis. Optional if
SRA accession is provided instead &lt;str&gt;</p>
</dd>
<dt>-o , –sample-output-dir</dt><dd><p>/path/to/sample &lt;str&gt;</p>
</dd>
<dt>-f , –fastq          Path to the fastq input file. Can be raw or with</dt><dd><p>adapters removed</p>
</dd>
<dt>-f1 , –fastq-r1      Path to the mate 1 fastq input file, if reads are PE.</dt><dd><p>Can be raw or with adapters removed</p>
</dd>
<dt>-f2 , –fastq-r2      Path to the mate 2 fastq input file, if reads are PE.</dt><dd><p>Can be raw or with adapters removed</p>
</dd>
<dt>-SA , –sra           Fetch raw data files from the SRA using the provided</dt><dd><p>accession code &lt;str&gt;</p>
</dd>
</dl>
<dl class="option-list">
<dt><kbd><span class="option">-C</span>, <span class="option">--collapse</span></kbd></dt>
<dd><p>Collapse paired end reads &lt;bool&gt; (default: False)</p>
</dd>
</dl>
<dl class="simple">
<dt>-T , –trim_adapters</dt><dd><p>Remove adapters from raw fastq files &lt;bool&gt; (default:
True)</p>
</dd>
<dt>-TF , –adaperremoval-flags</dt><dd><p>Additional flags to provide to Adapterremoval &lt;str&gt;</p>
</dd>
</dl>
<p>-c , –cores          Number of cores for RIP to use
-M , –mem            Max memory resources allowed to be used ofr indexing</p>
<blockquote>
<div><p>the input for the filtering alignment (default: max
available memory 8192.0)</p>
</div></blockquote>
<dl class="option-list">
<dt><kbd><span class="option">-u</span>, <span class="option">--unlock</span></kbd></dt>
<dd><p>Unlock the working directory after smk is abruptly
killed &lt;bool&gt; (default: False)</p>
</dd>
<dt><kbd><span class="option">-d</span>, <span class="option">--debug</span></kbd></dt>
<dd><p>Debug the RIP workflow &lt;bool&gt; (default: False)</p>
</dd>
</dl>
<p>-smk , –snakemake    Snakemake flags (default: ‘’)
–dry-run</p>
</div></blockquote>
</div>
<div class="section" id="rip-analyse">
<h2>rip analyse<a class="headerlink" href="#rip-analyse" title="Permalink to this headline">¶</a></h2>
<blockquote>
<div><dl class="option-list">
<dt><kbd><span class="option">-h</span>, <span class="option">--help</span></kbd></dt>
<dd><p>show this help message and exit</p>
</dd>
</dl>
<p>-m , –mode           Analysis mode for the selected sample
-D , –database       Path to the database used for the sample analysis.</p>
<blockquote>
<div><p>MANDATORY</p>
</div></blockquote>
<p>-S , –sample         Path to the sample parameter file. Optional
-g  [ …], –genera  [ …]</p>
<blockquote>
<div><p>List containing the names of specific genera the
abundances should be calculated on, separated by a
space character &lt;genus1 genus2 genus3 …&gt;</p>
</div></blockquote>
<dl class="simple">
<dt>-o , –analysis-output-dir</dt><dd><p>Path to results directory.</p>
</dd>
</dl>
<p>-c , –cores          Number of cores for RIP to use
-M , –mem            Max memory resources allowed to be used ofr indexing</p>
<blockquote>
<div><p>the input for the filtering alignment (default: max
available memory 8192.0)</p>
</div></blockquote>
<dl class="option-list">
<dt><kbd><span class="option">-u</span>, <span class="option">--unlock</span></kbd></dt>
<dd><p>Unlock the working directory after smk is abruptly
killed &lt;bool&gt; (default: False)</p>
</dd>
<dt><kbd><span class="option">-d</span>, <span class="option">--debug</span></kbd></dt>
<dd><p>Debug the RIP workflow &lt;bool&gt; (default: False)</p>
</dd>
</dl>
<p>-smk , –snakemake    Snakemake flags (default: ‘’)
–dry-run</p>
</div></blockquote>
</div>
</div>
<div class="section" id="citations">
<h1>Citations<a class="headerlink" href="#citations" title="Permalink to this headline">¶</a></h1>
<p>bowtie2
mapdamage
adaperremoval
snakemake
whatever else</p>
</div>
<div class="section" id="workflow">
<h1>Workflow<a class="headerlink" href="#workflow" title="Permalink to this headline">¶</a></h1>
<p>To be filled in</p>
</div>
<div class="section" id="faqs">
<h1>FAQs<a class="headerlink" href="#faqs" title="Permalink to this headline">¶</a></h1>
</div>
<div class="section" id="contributing">
<h1>Contributing<a class="headerlink" href="#contributing" title="Permalink to this headline">¶</a></h1>
<p>Evangelos Antonios Dimopoulos,
Evan K. Irving-Pease,
Alberto Carmagnini</p>
</div>
<div class="section" id="credits">
<h1>Credits<a class="headerlink" href="#credits" title="Permalink to this headline">¶</a></h1>
<p>Don’t know who to credit</p>
</div>
<div class="section" id="change-log">
<h1>Change Log<a class="headerlink" href="#change-log" title="Permalink to this headline">¶</a></h1>
</div>
<div class="section" id="license">
<h1>License<a class="headerlink" href="#license" title="Permalink to this headline">¶</a></h1>
<p>MIT</p>
</div>
<div class="section" id="indices-and-tables">
<h1>Indices and tables<a class="headerlink" href="#indices-and-tables" title="Permalink to this headline">¶</a></h1>
<ul class="simple">
<li><p><a class="reference internal" href="genindex.html"><span class="std std-ref">Index</span></a></p></li>
<li><p><a class="reference internal" href="py-modindex.html"><span class="std std-ref">Module Index</span></a></p></li>
<li><p><a class="reference internal" href="search.html"><span class="std std-ref">Search Page</span></a></p></li>
</ul>
</div>


          </div>
          
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
<h1 class="logo"><a href="#">RIP</a></h1>








<h3>Navigation</h3>

<div class="relations">
<h3>Related Topics</h3>
<ul>
  <li><a href="#">Documentation overview</a><ul>
  </ul></li>
</ul>
</div>
<div id="searchbox" style="display: none" role="search">
  <h3 id="searchlabel">Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" aria-labelledby="searchlabel" />
      <input type="submit" value="Go" />
    </form>
    </div>
</div>
<script>$('#searchbox').show(0);</script>








        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="footer">
      &copy;2020, Evangelos A. Dimopoulos, Evan K. Irving-Pease.
      
      |
      Powered by <a href="http://sphinx-doc.org/">Sphinx 3.1.2</a>
      &amp; <a href="https://github.com/bitprophet/alabaster">Alabaster 0.7.12</a>
      
      |
      <a href="_sources/index.rst.txt"
          rel="nofollow">Page source</a>
    </div>

    

    
  </body>
</html>