HAYSTAC documentation
===============================

.. toctree::
   :maxdepth: 2
   :caption: Contents:

   introduction
   installation
   workflow
   tutorial
   cli
   developer


Introduction
============

HAYSTAC is a comprehensive computational tool for identifying species from DNA sequence data. It can pre-process sequencing data (adapter trimming), build a database, analyse sequencing data and perform a deamination profile analysis. It can work in two different modes. One mode performs metagenomic identifications from samples containing multiple organisms, and outputs mean posterior species abundances. The second mode can perform species identifications from single organism samples and it outputs species assignment posterior probabilities.


Citations
========= 

bowtie2: https://doi.org/10.1038/nmeth.1923
mapdamage2: https://doi.org/10.1093/bioinformatics/btt193
adapterremoval: https://doi.org/10.1186/s13104-016-1900-2
snakemake: https://doi.org/10.1093/bioinformatics/bts480


FAQs 
====


Contributing
============
Evangelos Antonios Dimopoulos,
Evan K. Irving-Pease,
Alberto Carmagnini


Change Log 
==========


License
=======

MIT


